### PR TITLE
mod data

### LINKED ISSUE #65: sort mods
like ghost categories, lets identify armor mods by
• purpose ("load", "ammo finder", "unflinching")
• enhancedness
• weapon affinity

this isn't hugely useful but would mean we can sort/group mods page, mods socket popup, and potentially loadout builder options, in a far more useful way than d2.exe currently does

### PR DESCRIPTION
**Pull request by delphiactual**
Generate some mod data to start with.

Closes #65

### CODE REVIEW

File: generate-mod-data.js [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    item.perks.length
+  ) {
+    const name = item.displayProperties.name;
+    const description = perks[item.perks[0].perkHash].displayProperties.description || '';
+    const hash = item.hash;
+    const enhanced = getEnhanced(name);
+    const stacks = getStacks(description);
+    const affinity = (item.plug.energyCost && item.plug.energyCost.energyType) || 0;
+
+    for (i = 0; i < item.plug.insertionRules.length; i++) {
```
Comment (delphiactual) [inline generate-mod-data.js]: found up to 4 failureMessages

File: generate-mod-data.js [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+}
+
+function getFailureMessages(item) {
+  for (i = 0; i < item.plug.insertionRules.length; i++) {
+    if (failureMessages[item.plug.insertionRules[i].failureMessage]) {
+      failureMessages[item.plug.insertionRules[i].failureMessage].push(hash);
+    } else {
+      failureMessages[item.plug.insertionRules[i].failureMessage] = [hash];
+    }
+  }
```
Comment (delphiactual) [inline generate-mod-data.js]: found up to 4 failureMessages per mod

File: generate-mod-data.js [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+}
+
+function getFailureMessages(item) {
+  for (i = 0; i < item.plug.insertionRules.length; i++) {
+    if (failureMessages[item.plug.insertionRules[i].failureMessage]) {
+      failureMessages[item.plug.insertionRules[i].failureMessage].push(item.hash);
+    } else {
+      failureMessages[item.plug.insertionRules[i].failureMessage] = [item.hash];
+    }
+  }
```
Comment (delphiactual) [inline generate-mod-data.js]: found up to 4 failureMessage per mod, https://data.destinysets.com/i/InventoryItem:3175015261

Comment (delphiactual) [reply generate-mod-data.js]: we might want to know where each failureMessage is at per hash.

**Comment by delphiactual** [comment]
Possible consider all properties optional and only output if they are outside of default values
e.g. enhanced > 0, affinity > 0, stacks === true

**Comment by delphiactual** [comment]
@sundevour what do you think about maybe re-org the `mod-data-failure-messages.json` to be `[index][failureMessage][hash]`?

File: generate-mod-data.js [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+
+const DEBUG = false;
+
+const mods = {};
+const failureMessages = {};
+
+Object.values(inventoryItems).forEach((item) => {
+  if (
+    item.itemCategoryHashes &&
+    item.itemCategoryHashes.includes(59) &&
```
Comment (delphiactual) [inline generate-mod-data.js]: Change to just Armor Mods?

Comment (delphiactual) [reply generate-mod-data.js]: reduces mod data to 827 lines

File: generate-mod-data.js [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+  }
+  if (name.includes('Energy Weapon')) {
+    ich.push(itemCategoryHash.ENERGY);
+  }
+  if (name.includes('Kinetic Weapon')) {
+    ich.push(itemCategoryHash.KINETIC);
+  }
+  return ich;
+}
+function getFailureMessages(item) {
```
Comment (delphiactual) [inline generate-mod-data.js]: Remove

File: generate-mod-metadata.js [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+    ich.push(itemCategoryHash.TRACE_RIFLE);
+  }
+  if (name.includes('Bow')) {
+    ich.push(itemCategoryHash.BOW);
+  }
+  if (name.includes('Submachine Gun')) {
+    ich.push(itemCategoryHash.SMG);
+  }
+  if (name.includes('Precision Weapon')) {
+    // and slug shotguns
```
Comment (delphiactual) [inline generate-mod-metadata.js]: how do I figure out if a shotgun is a slug shotgun?

**Comment by sundevour** [comment]
are we enriching affinity in any way?
or are there DIM circumstances where we'd have a mod but not its definition available?

**Comment by delphiactual** [comment]
nope i am an idiot, and just read `affinity` in #65

**Pull request closed without merging**